### ISSUE TITLE
strange error in some cases

### ISSUE THREAD

**Issue by zoff99** [sksamuel/scrimage#293]
something broke, i am not sure what

```
shrink_image_file:compressToFile:EE003:Image parsing failed for PNG. If the format is `webp` ensure you have a webp reader on your classpath, such as the `scrimage-webp` module. Tried the following ImageReader implementations:
com.sksamuel.scrimage.webp.WebpImageReader@37cc846a failed due to java.io.IOException: [Decoding of /tmp/input16582128566075645218webp failed., Status: 3(BITSTREAM_ERROR)]
com.sksamuel.scrimage.ImageParseException: Image parsing failed for PNG. If the format is `webp` ensure you have a webp reader on your classpath, such as the `scrimage-webp` module. Tried the following ImageReader implementations:
com.sksamuel.scrimage.webp.WebpImageReader@37cc846a failed due to java.io.IOException: [Decoding of /tmp/input16582128566075645218webp failed., Status: 3(BITSTREAM_ERROR)]
	at com.sksamuel.scrimage.nio.ImageReaders.read(ImageReaders.java:66)
	at com.sksamuel.scrimage.nio.ImageReaders.read(ImageReaders.java:37)
	at com.sksamuel.scrimage.nio.ImmutableImageLoader.load(ImmutableImageLoader.java:139)
	at com.sksamuel.scrimage.nio.ImmutableImageLoader.fromFile(ImmutableImageLoader.java:106)
	at com.zoffcc.applications.trifa.HelperGeneric.shrink_image_file(HelperGeneric.kt:366)
	at com.zoffcc.applications.trifa.MainActivity$Companion.add_ngc_outgoing_file(MainActivity.kt:3788)
	at ChatAppKt$GroupApp$1$1.invoke$lambda$24$lambda$23$lambda$11$lambda$10(ChatApp.kt:467)
	at androidx.compose.ui.AwtWindowDropTarget$dragTargetListener$4.invoke(ExternalDrag.desktop.kt:243)
	at androidx.compose.ui.AwtWindowDropTarget$dragTargetListener$4.invoke(ExternalDrag.desktop.kt:233)
	at androidx.compose.ui.AwtWindowDragTargetListener.drop(ExternalDrag.desktop.kt:450)
	at java.desktop/java.awt.dnd.DropTarget.drop(DropTarget.java:455)
	at java.desktop/sun.awt.dnd.SunDropTargetContextPeer.processDropMessage(SunDropTargetContextPeer.java:548)
	at java.desktop/sun.awt.X11.XDropTargetContextPeer.processDropMessage(XDropTargetContextPeer.java:185)
	at java.desktop/sun.awt.dnd.SunDropTargetContextPeer$EventDispatcher.dispatchDropEvent(SunDropTargetContextPeer.java:864)
	at java.desktop/sun.awt.dnd.SunDropTargetContextPeer$EventDispatcher.dispatchEvent(SunDropTargetContextPeer.java:788)
	at java.desktop/sun.awt.dnd.SunDropTargetEvent.dispatch(SunDropTargetEvent.java:48)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4861)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2780)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4828)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:775)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:720)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:714)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:98)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:747)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:744)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

```

**Comment by sksamuel** [comment]
Do you have the source image?

**Comment by zoff99** [comment]
its not the image, for some reason when i run that multiplatform app via "./gradlew run" it works and when i package it it also works. but when i run the java commandline freehand it produces this error.
i don't know really where to begin looking

**Comment by sonianvs** [comment]
Hi there, 

I've a similar issue after adding the `scrimage-webp` module to our dependencies, which starts to be frustrating as I can't find the origin, but I'll explain what I've found in case it helps (and you can help me, 🙏🏻 ):

**Before**: 
We had both `scrimage-core` and `scrimage-formats-extra` dependencies and we are loading images with the following line of code:  `ImmutableImage.loader.fromStream(is)`. To put you more in context, I'm using Scala language and I run it locally with sbt, but I package it to a jar to deploy the code in an AWS lambda. 

**Change made**: 
We wanted to add the `scrimage-webp` dependency so we could support that kind of image files. So I added that module to the dependencies.

**After**: 

I have 2 different behaviours: 

**1**. If I add the `scrimage-webp` dependency **on top of the list** (before the other two mentioned above): 

- Locally: I can run it and process any kind of image (including webp) with no issues. 

- Packaged and deployed to an AWS Lambda: 

I can upload webp image files and it runs successfully but then I have the error with jpg files:

<img width="1274" alt="Captura de pantalla 2024-09-16 a les 13 43 57" src="https://github.com/user-attachments/assets/773eceb7-2c71-4773-a4ec-c10a26ef224f">


**2**. If I add the `scrimage-webp` dependency **at the bottom of the list** (after the other two mentioned above): 

- Locally: I can run it and process any kind of image (including webp) with no issues. 

- Packaged and deployed to an AWS Lambda: 

I can upload any kind of image files and it runs successfully but then I have the error with webp files (as before adding the webp module):

<img width="1243" alt="Captura de pantalla 2024-09-16 a les 13 54 39" src="https://github.com/user-attachments/assets/c9f42475-8583-4b35-b3d8-98cb915358b0">

It looks like a dependency issue, but I can't find the origin. For more details: I'm using the latest versions for the scrimage dependencies (4.2.0) and the runtime is Java 11.  

Any ideas?

**Comment by zoff99** [comment]
@sksamuel can we debug this together please? this would be really great.

**Comment by zoff99** [comment]
i have made a minimal KMP example to show the issue.
can you check it out please?

```
git clone https://github.com/zoff99/testbed_debug
TERM=dumb ./gradlew run
# input file exists?: true <-- it logs that the input file exists
# CTRL-C
```

it will happily use `build/compose/tmp/prepareAppResources/background.jpg` and create `/tmp/aa.webp`

now when you package it and run it with java

```
./manual_run.sh build
# input file exists?: true <-- it logs that the input file exists
```

you can see it has the exception. but i can not figure out why it happens. the resouces path and the input file are there.
can you please have a look?

thank you.

**Comment by zoff99** [comment]
@sksamuel hope you are still around

**Comment by sksamuel** [comment]
Thanks for the repo, that helped.
The issue is you're not merging the service files when you build a fat jar. This prob breaks other  libraries too.

I've updated the FAQ page with details on how to fix:
https://sksamuel.github.io/scrimage/faq/

**Comment by zoff99** [comment]
the gradle link on that page is a 404
https://gradleup.com/shadow/configuration/merging-services/

i am not sure thats really the issue. what i have done is looking at the "ps" output when running "./gradlew run" and then just running the "java" command that i see without gradle. that must do the exact same thing, shouldnt it?

all other stuff is working fine.

**Comment by sksamuel** [comment]
I'm pretty sure this is your issue. That's why flipping the dependency order changed which one is picked up

**Comment by zoff99** [comment]
for context, the reason i am running it this way, is that i need (want) to run the native JNI libs with ASAN. and i have found no other way this works. with "gradlew run" it wont even start up.

**Comment by zoff99** [comment]
can you fix the link on the https://sksamuel.github.io/scrimage/faq/ page?
it points to (after some redirects) https://gradleup.com/shadow/configuration/merging-services/
which seems to be a 404
not sure where it should actually point to.

**Comment by sksamuel** [comment]
you're using scala you said ?
https://github.com/sbt/sbt-assembly?tab=readme-ov-file#merge-strategy

**Comment by zoff99** [comment]
no. using gradle

**Comment by sksamuel** [comment]
https://gradleup.com/shadow/configuration/merging/#merging-service-descriptor-files

**Comment by sonianvs** [comment]
Thanks a lot @sksamuel for your help! As you suggested, the origin of the issue in my case was due to how the `assemblyMergeStrategy` was implemented. 

Just in case someone come across this thread having the same issue and using scala: 

We had and `assemblyMergeStrategy` defined with a pattern matching using `MergeStrategy.first` for default cases as a workaround since `MergeStrategy.deduplicate` is not working for us (due to some issues in our dependency graph we need to work on) and, thus, we cannot use the `defaultMergeStrategy` in sbt-assembly. We hadn't a case defined in our pattern matching for the service files and, therefore, were falling in that default case. 

I've reimplemented the assemblyMergeStrategy this way: 

```
    assembly / assemblyMergeStrategy := { file =>
        val oldStrategy = (assembly / assemblyMergeStrategy).value
        if (oldStrategy(file).name == "Deduplicate") MergeStrategy.first else oldStrategy(file)
      }
```

This way, we are using the defaultMergeStrategy except by when the `deduplicate` strategy would be applied in which case we decide to apply the `first` strategy (for now, while we work on our dependencies). It is not ideal, it is only a workaround. As stated in the sbt-assembly readme: 

> The real solution is to fix that dependency graph.

Now, the webp is added to the services folder in the jar packaging and the images are being correctly processed.
